tools/gfx/vulkan/render-vk.cpp

```diff
@@ -879,322 +676,2084 @@ class VKDevice : public GraphicsAPIRenderer
+        ///
+        /// Created on demand with `_createOrdinaryDataBufferIfNeeded()`
+        ComPtr<IBufferResource> m_ordinaryDataBuffer;
+
+        /// Get the layout of this shader object with specialization arguments considered
+        ///
+        /// This operation should only be called after the shader object has been
+        /// fully filled in and finalized.
+        ///
+        Result _getSpecializedLayout(ShaderObjectLayoutImpl** outLayout)
+        {
+            if (!m_specializedLayout)
+            {
+                SLANG_RETURN_ON_FAIL(_createSpecializedLayout(m_specializedLayout.writeRef()));
+            }
+            *outLayout = RefPtr<ShaderObjectLayoutImpl>(m_specializedLayout).detach();
+            return SLANG_OK;
+        }
+
+        /// Create the layout for this shader object with specialization arguments considered
+        ///
+        /// This operation is virtual so that it can be customized by `ProgramVars`.
+        ///
+        virtual Result _createSpecializedLayout(ShaderObjectLayoutImpl** outLayout)
+        {
+            ExtendedShaderObjectType extendedType;
+            SLANG_RETURN_ON_FAIL(getSpecializedShaderObjectType(&extendedType));
+
+            auto renderer = getRenderer();
+            RefPtr<ShaderObjectLayoutBase> layout;
+            SLANG_RETURN_ON_FAIL(
+                renderer->getShaderObjectLayout(extendedType.slangType, layout.writeRef()));
+
+            *outLayout = static_cast<ShaderObjectLayoutImpl*>(layout.detach());
+            return SLANG_OK;
+        }
+
+        RefPtr<ShaderObjectLayoutImpl> m_specializedLayout;
+    };
+
+    class EntryPointShaderObject : public ShaderObjectImpl
+    {
+        typedef ShaderObjectImpl Super;
+
+    public:
+        static Result create(
+            IDevice* device,
+            EntryPointLayout* layout,
+            EntryPointShaderObject** outShaderObject)
+        {
+            RefPtr<EntryPointShaderObject> object = new EntryPointShaderObject();
+            SLANG_RETURN_ON_FAIL(object->init(device, layout));
+
+            *outShaderObject = object.detach();
+            return SLANG_OK;
+        }
+
+        EntryPointLayout* getLayout() { return static_cast<EntryPointLayout*>(m_layout.Ptr()); }
+
+        virtual Result bindObject(
+            PipelineCommandEncoder* encoder,
+            RootBindingState* bindingState,
+            BindingOffset& offset) override
+        {
+            // Copy entry point uniform data into global push constant buffer.
+            if (m_ordinaryData.getCount())
+            {
+                Index dstOffset = bindingState->pushConstantBuffer.getCount();
+                ::memcpy(
+                    bindingState->pushConstantBuffer.getBuffer() + dstOffset,
+                    m_ordinaryData.getBuffer(),
+                    m_ordinaryData.getCount());
+                bindingState->pushConstantBuffer.setCount(dstOffset + m_ordinaryData.getCount());
+            }
```

Comment:
We've discussed it in some of these other PRs, so I won't belabor it here, but it seems weird to me to copy all the uniform/ordinary data from all the entry points into a distinct buffer so that we can make a single VK call to write the push constants instead of just making the calls right here. We have the data, and we know the offset where we want to write it, so why not just make the right VK API call here?